File: lab3/task4.cpp
#include <iostream>      
#include <vector>      
#include <algorithm>    
#include <climits>       
using namespace std;     

int main() {
    int n, m;            // n — сколько чисел, m — максимум за ход
    cin >> n >> m;       // читаем n и m
    
    vector<int> a(n);    // массив чисел
    for (int i = 0; i < n; ++i) {
        cin >> a[i];     // заполняем массив
    }
    
    vector<int> pref(n + 1, 0);   // pref[0] = 0 ref[i] = сумма первых i элементов
    for (int i = 0; i < n; ++i) {
        pref[i + 1] = pref[i] + a[i];   // считаем суммы
    }
    
    int scoreP = 0, scoreV = 0;   // счёт Павла и Вики
    int start = 0;                // с какого индекса сейчас играем
    int last_move = -1;           // сколько взял противник в прошлый раз (-1 = нет прошлого)
    bool isPavel = true;          // true = ход Павла, false = ход Вики
    while (start < n) {           // пока ещё есть числа
        int best_sum = INT_MIN;   // лучшая сумма за ход (пока минимальная)
        int best_k = 1;           // лучшее количество чисел (пока 1)
        
        int max_possible = min(m, n - start);   // сколько максимум можно взять 
        for (int k = 1; k <= max_possible; ++k) {   // пробуем k от 1 до max
            if (k == last_move) continue;   // нельзя брать столько же, сколько противник
            int sum = pref[start + k] - pref[start];   // сумма k чисел от start
    
            if (sum > best_sum || (sum == best_sum && k < best_k)) {   // если сумма больше лучшей, или такая же но k меньше — выбираем этот k
                best_sum = sum;
                best_k = k;
            }
        }

        if (isPavel) {      // добавляем лучшую сумму к счёту текущего игрока
            scoreP += best_sum;
        } else {
            scoreV += best_sum;
        }  
        start += best_k;          // сдвигаем начало
        last_move = best_k;       // запоминаем, сколько взяли
        isPavel = !isPavel;       // передаём ход другому
    }
    cout << (scoreP > scoreV ? 1 : 0) << endl;    // 1 — победил Павел, 0 — победила Вика
    return 0;                    
}

File: lab3/task4.go
package main

import "fmt"

func main() {
	var n, m int
	fmt.Scan(&n, &m)      // считываем количество чисел и максимум за ход
	a := make([]int64, n) // Создаём массив чисел
	for i := 0; i < n; i++ {
		fmt.Scan(&a[i])
	}

	// Префиксные суммы — чтобы быстро считать сумму любого отрезка
	pref := make([]int64, n+1)
	for i := 0; i < n; i++ {
		pref[i+1] = pref[i] + a[i]
	}

	var scoreP, scoreV int64 // счета Павла и Вики
	start := 0               // индекс первого нестёртого числа
	lastMove := -1           // сколько чисел взял противник в прошлый раз (-1 = ходов ещё не было)
	isPavel := true          // Павел ходит первым
	for start < n {
		bestSum := int64(-1 << 62) //Начинаем с минимальной суммы, чтобы любая нормальная сумма была больше
		bestK := 1

		maxK := m // Нельзя взять больше, чем осталось, и больше m
		if n-start < maxK {
			maxK = n - start
		}

		// Перебираем все возможные варианты взятия k чисел
		for k := 1; k <= maxK; k++ {
			// Ключевое условие варианта 2: нельзя брать столько же, сколько взял противник в прошлый раз
			if k == lastMove {
				continue // запрещённый ход, пропускаем
			}
			sum := pref[start+k] - pref[start]

			// Выбираем k с максимальной суммой
			// Если суммы равны, берём меньшее k (по условию игры)
			if sum > bestSum || (sum == bestSum && k < bestK) {
				bestSum = sum
				bestK = k
			}
		}

		// Добавляем выбранную сумму к счёту текущего игрока
		if isPavel {
			scoreP += bestSum
		} else {
			scoreV += bestSum
		}

		// Сдвигаем начало массива, запоминаем, сколько взяли, передаём ход
		start += bestK
		lastMove = bestK
		isPavel = !isPavel
	}

	if scoreP > scoreV { // 1-пашка 0-викуля - чикуля
		fmt.Println(1)
	} else {
		fmt.Println(0)
	}
}
